snapcraft/internal/sources/_base.py

```diff
@@ -45,12 +46,43 @@ def __init__(self, source, source_dir, source_tag=None, source_commit=None,
 class FileBase(Base):
 
     def pull(self):
-        if snapcraft.internal.common.isurl(self.source):
-            self.download()
-        else:
-            shutil.copy2(self.source, self.source_dir)
+        file_cache = FileCache()
+        source_file = None
+        is_source_url = snapcraft.internal.common.isurl(self.source)
+
+        # First check if we already have the source file cached.
+        if self.source_checksum:
+            algorithm, hash = split_checksum(self.source_checksum)
+            cache_file = file_cache.get(algorithm=algorithm, hash=hash)
+            if cache_file:
+                source_file = os.path.join(self.source_dir,
+                                           os.path.basename(cache_file))
+                # We make this copy as the provisioning logic can delete
+                # this file and we don't want that.
+                shutil.copy2(cache_file, source_file)
 
-        self.provision(self.source_dir)
+        # If not, first check if it is a url and download and if not
+        # it is probably locally referenced.
+        if not source_file and is_source_url:
+            source_file = self.download()
+        elif not source_file:
+            basename = os.path.basename(self.source)
+            source_file = os.path.join(self.source_dir, basename)
+            # We make this copy as the provisioning logic can delete
+            # this file and we don't want that.
+            shutil.copy2(self.source, source_file)
+
+        # Before provisioning we verify the file if source_checksum is defined
+        # and we cache the file for future reuse.
+        if self.source_checksum:
+            algorithm, digest = verify_checksum(
+                self.source_checksum, source_file)
+            file_cache.cache(filename=source_file,
+                             algorithm=algorithm,
+                             hash=hash)
```

Comment:
ah... I want to keep hash and algorithm separate from how it is represented in a source file. Maybe the better way would be to split at `__init__`.
This is because this is not the only place we check for hashes, what is sadder is that the implementation that does the checksum here does it wrong as it does not chunk it correctly :-/